discovery-handlers/udev/src/discovery_handler.rs

```diff
@@ -68,37 +71,64 @@ impl DiscoveryHandler for DiscoveryHandlerImpl {
                     }
                     break;
                 }
-                let mut devpaths: HashSet<DeviceProperties> = HashSet::new();
+                let mut devpaths: HashMap<String, Vec<DeviceProperties>> = HashMap::new();
                 udev_rules.iter().for_each(|rule| {
                     let enumerator = udev_enumerator::create_enumerator();
                     let paths = do_parse_and_find(enumerator, rule).unwrap();
-                    paths.into_iter().for_each(|path| {
-                        devpaths.insert(path);
-                    });
+                    for path in paths.into_iter() {
+                        if !discovery_handler_config.group_recursive {
+                            devpaths.insert(path.0.clone(), vec![path]);
+                        } else {
+                            match get_device_relatives(&path.0, devpaths.keys()) {
```

Comment:
nit: i would put this else block in a separate function so we can unit test it.

Reply:
will do this next week, as I'll do the unit tests as well (won't have enough time before that), if you want to merge before that, I'll do it in a separate PR.